### ISSUE TITLE
bug: Ionic View transition flickering

### ISSUE THREAD

**Issue by svsool** [driftyco/ionic#4021]
**Type**: <span ionic-type>bug</span>

**Platform**: <span ionic-platform>ios</span> <span ionic-platform-version>8</span>  <span ionic-webview>webview</span>

<span ionic-description>There are problem with view transition in latest version, leaving view takes half display and overlap enter view, here bug localization
View animations in 1.0.0-beta.13, is ok, animations smooth
http://plnkr.co/edit/jxnbgCABggbqTTOwGPIp?p=preview
View animations in 1.0.1 flickering http://plnkr.co/edit/1LYhSNLo5fxzMg1uzVGY?p=preview, many stage and leaving views created in dom on transition with style translate3d(33%, 0, 0) and not removed (http://joxi.ru/gV2VRBdiBdDv2v.jpg) 
On my ipad mini, 1.0.0-beta.13 working much faster than 1.0.1</span>

<span is-issue-template></span>

**Comment by Ionitron** [comment]
Greetings @svsool!

My sensors indicate that you need to update your issue through our [custom issue form](http://ionicframework.com/submit-issue/#/?iid=4021). We are now requiring all issues to be submitted this way, to ensure that we have all of the information necessary to fix them as quickly as possible.

**[Click Here To Update Your Issue](http://ionicframework.com/submit-issue/#/?iid=4021)**

I will have no choice but to close this issue if it is not resubmitted through the form. Please fill out the rest of the form, so that I may use my friendly robot powers to assist you.

Thank you!

<span ionitron-issue-resubmit></span>

**Comment by svsool** [comment]
for localize bug example (sidebarapp), only background .scroll element changed, all other native

**Comment by foxgeere** [comment]
Can confirm have this issue on multiple iPhones as well here with ionic lib 1.0.1. Thought we had broken something. Any temp fix welcome but hopefully fixed in next release. :)

**Comment by foxgeere** [comment]
Still having this issue on 1.1.0 fyi

**Comment by mhartington** [comment]
So there is a lot of noise in the plunkr, it a bit hard to illustrate the issues. The constant transition back and forth doesn't really help...

So the change you are talking about form beta 13 to nightly are around the navigations style, which are much more closes to iOS. Pulled do the plunkr with the latest version of ionic and loaded it on an iphone and nexus 5. Wasn't getting any flickers using that test. 

If you could provide a more clear demo on how I could recreate the issue, that would be great.

**Comment by foxgeere** [comment]
Stand by. Will try and screen record it from the iOS simulator.

**Comment by foxgeere** [comment]
Simulator has the power of the mac and is just a touch to quick to properly illustrate what I get on the iPhone 6 (and worse on older devices).

Ultimately it's as if two copies of the view you are 'leaving' slide over the view you are entering. One goes all the way to the left (as it should) the other only moves about 1/4 of the way to the left then freezes until the end of the transition, then the remain 3/4 vanish.

This didn't use to happen for sure as I am working on the UI aspects while others are dealing with the controllers etc. It makes it quite jerky / strange. It isn't a show stopped so we have just left it on this bug report for the time being.

The second plunkr above does demo the issue but could do with some delays put in so it doesn't keep repeating so quickly, or could be triggered manually. I am unaware of how to edit it for that purpose.

I tried to get a screen shot. What you see is a full width of an iPhone screen (I have had to crop top and bottom to avoid leaking info about the app ;) ).

You can see that one copy of the old view has slid over while another copy is stuck 1/4 of the way across. This happens on every view where we transition from a list to a sub view.

I could perhaps share the app privately via ionic view.

![screen shot 2015-08-19 at 17 34 59](https://cloud.githubusercontent.com/assets/10975935/9362688/f3592f4e-4699-11e5-99ae-a82e8846a63f.png)

As an addition to this, when we transition into the view we also notice that the nav bar icons show as 'grey blobs' during the transition and only become the final ion-icon once the transition is complete like this:

![img_0042](https://cloud.githubusercontent.com/assets/10975935/9362778/728ed750-469a-11e5-977b-a07300083988.jpg)

becomes this after transition complete

![img_0043](https://cloud.githubusercontent.com/assets/10975935/9362784/7d7b01c0-469a-11e5-807d-15ce33833ca0.jpg)

**Comment by mhartington** [comment]
So I tested the animations and compared them to a native app.

Ionic's
https://youtu.be/Vl-j_8-1qBY

Native
https://youtu.be/UKltEjyJdWk

Both animations are practically identical. 
So I'm thinking the flickering or issues with the animations are from code you have added
Can you start slowly removing some css/js and see where the bottleneck is?

**Comment by svsool** [comment]
For my app I use old version of ionic, because sometimes after manually triggering route transition  i get flickering (for my particularly case app while disconnected from remote hardware will transition to connect page and if connection very fast return back on previouse page without delay and in this moment i get flickering), for debuging this case, with canary i toggle animation delay to 0.1x in devtools and notice that many in/out layout generated by ionic. I think it's problem with rendering queue, in corner version if ionic used css animations instead of angular animation, with angular animation perfomance is ok, with css animations there are some troubles with rendering queue, if transitions really fast between pages it will be seen

**Comment by mhartington** [comment]
@svsool would you be able to email me? It would be good to know the specifics of this to help resolve the issue. 

mike at ionic dot io

**Comment by carloslema** [comment]
@svsool & @mhartington, hi guys. 

I got the same transition issue that svsool is talking about. Started happening recently, maybe as of last week.

Went back to ionic 1.0.0, as suggested and I am still getting that jump. It only happens going from a Parent List > to Child List. Parent List's State Cache is disabled, but not on the Child.

I am testing on iPhone 4 iOS 7.1, iPhone 5 iOS 8.4, same thing. BUT, on Galaxy S4 Android 5.1. the transition jump does not show up.

Hope this helps finding the issue.
\\ carlos

**Comment by mhartington** [comment]
@carloslema your case, it's probably due to the cache being disabled.

**Comment by carloslema** [comment]
@mhartington just tried. nope. What is puzzling, why on Android works fine but iOS has that overshooting hiccup. By the way, @mhartington have you been able to recreate this on your end?

**Comment by mhartington** [comment]
No I have not, I posted a video above and the animations are smooth for me, and that was on an old iphone 5

**Comment by carloslema** [comment]
@mhartington here we go, http://youtu.be/ChbBXPiZM-M -- didn't find a ios video screen recorder

**Comment by carloslema** [comment]
@mhartington following up with you guys, I updated the video, this time is iPhone 5, http://youtu.be/ErBLb0l5H3g

The sequence: After I click on the list item..
-- "From window" moves to the left, it looks like it is masked. A black bar jumps in 37px from left to right (seams to be a background). 
-- The "To window" continues to slide in from the right, in between the Tabs & the "from window" 
-- the black bar is moving to the left.
-- When the "To window" passes the half way point, the "To window" jumps back to the right. 
-- Tabs move back to place, the "From window" & back bar slide out, as the "To window" slides in.
-- Transition animation is completed. This happens in about 17 frames.

ng-cloak is also not working, but that is another issue.

![ionic-transition](https://cloud.githubusercontent.com/assets/10660649/9425464/6722adfa-48df-11e5-8ae0-2a25ddd14ea2.jpg)

This happens on both, iOS Simulators and iPhone 4 & 5, iPad 2.
Also, as I said before, this does not show up on an actual Android Phone or the Android Simulator.

Your system information:
Cordova CLI: 5.1.1
Gulp version:  CLI version 3.8.11
Gulp local:   Local version 3.9.0
Ionic Version: 1.1.0
Ionic CLI Version: 1.6.4
Ionic App Lib Version: 0.3.8
ios-deploy version: 1.7.0 
ios-sim version: 3.1.1 
OS: Mac OS X Yosemite
Node Version: v0.12.7
Xcode version: Xcode 6.4 Build version 6E35b

**Comment by mhartington** [comment]
So there are a few obvious things that stand out there.
1) ion-view shouldn't be animating over the ion-tabs. What does your structure look like for this?
2) the animations aren't correct. Are you using any custom animations or have some css that could be interfering in anyway?

The video raises more questions if anything and it's not clear what the issue is just from watching. You'd need to go through the code, figure out if everything is setup correctly, checkout any conflicting css, etc. 

Let me know if you find anything.

**Comment by carloslema** [comment]
Thanks @mhartington for the follow up.

So, I found the issue. Nothing to do with the set up of the Tabs (which resolves an ID before loading the view) or CSS animation (which I had none on this view).

I had a 'Hidden' Div (alternative UI to add a new List Item, temporary solution, back to a modal, yikes!):

<ion-view view-title="{{ 'sportTitle' | translate}}" class="app_genBkg">
    <ion-nav-buttons side="right">
        <button class="button button-clear" ng-click="toggle()">
            <i class="icon ion-android-add"></i>
        </button>
    </ion-nav-buttons>
    <ion-content>
        <ion-list>

            
            <ion-item class="item-icon-left" ng-repeat="task in sportCheckList" type="item-text-wrap" >
                <i class="icon checkbox">
                    <input type="checkbox" ng-click="completedTask($index)" ng-model="task.checked">
                </i>
                <h2>{{task.title}}</h2>
                <ion-option-button class="button-assertive" ng-click="removeItem(task)">
                    <i class="icon ion-ios-trash"></i>
                </ion-option-button>
            </ion-item>

           ** 
            
            <div class="item item-icon-right" ng-class="{'hidden' : visible}" ng-cloak>
                <input type="text" placeholder="Insert new item" ng-model="item.title" autofocus>
                <div class="icon ion-ios-checkmark-outline balanced" ng-click="saveToChecklist(item)"></div>
            </div>**

        </ion-list>
    </ion-content>
</ion-view>

Thanks for a kick @ss product ladies & gents!

**Comment by mhartington** [comment]
:+1:  Happy to hear that you got it resolved. 

Now for anyone else on this thread. Have you been able to resolve your issues? 
Itching to close this one.... :grin:

**Comment by foxgeere** [comment]
Not been able to resolve yet. Am trying stripping out different DIVs etc based on the above comments but nothing results in a fix as yet. Will aim to update again within 24 hours.

**Comment by foxgeere** [comment]
Right - so, appears the other day that ionic lib update wasn't actually updating things. I've removed everything, set up everything fresh from scratch and re-added the iOS platform. Now, running on the device with NO changes to the templates or code, the flicker is gone. So whatever issue was introduced appears to have been fixed in the latest version as far as I can see.

However, I am still seeing the same issue with the menubar icons being a grey blob / button until the view is fully in view before it then turns into the expected ionic-icon. Perhaps a separate issue with regard to menubar rendering?

**Comment by foxgeere** [comment]
Current ionic info reads as:

Cordova CLI: 5.2.0
Ionic Version: 1.1.0
Ionic CLI Version: 1.6.4
Ionic App Lib Version: 0.3.8
ios-deploy version: 1.5.0
ios-sim version: 3.1.1
OS: Mac OS X Yosemite
Node Version: v0.12.6
Xcode version: Xcode 6.4 Build version 6E35b

**Comment by mhartington** [comment]
Please open a separate issue.

**Comment by carloslema** [comment]
In the Spirit of collaboration, this is how I ended up solving my issue:

Approach 1.
Ionic’s Modal, which is super well documented. I found it to be a little overkill for what I needed.

Approach 2.
1. In the View I added < div dynamic="html" >< /div >:
2. In my controller:

    $scope.html = ’’;

-- The actual function call:
    $scope.toggle = function() {
    $scope.html =   ’< div class="item item-icon-right" >’ +
                    ’    < input type="text" placeholder="Insert new item" ng-model="item.title" autofocus >’ +
                    ’    < div class="icon ion-ios-checkmark-outline balanced" ng-click="saveToChecklist(item)" >< /div >’ +
                    ’< /div >’;
    }


-- The ng-click does it thing:
    $scope.saveToChecklist = function (item){
        $scope.sportCheckList.push(item);
	// save factory call..
        $scope.html = ’’; // and remove from view!
    }

3. Added to my directives:
    .directive( ’dynamic’, function ( $compile ) {
      return {
        restrict: ’A’,
        replace: true,
        link: function ( scope, ele, attrs ) {
          scope.$watch( attrs.dynamic, function( html ) {
            ele.html( html );
            $compile( ele.contents() )( scope );
          });
        }
      };
    })

The view does not render or hide a DIV, which caused the Jump between views.
From an user experience, IMHO approach 2 is a little more transparent -- ng-blur was a little temperamental.
I am sure there are more elegant ways, but my caffeine is running low plus, got to put some miles on da bike.

Hope this helps, but If anyone has a better solution, I am all ears!

**Comment by mhartington** [comment]
@carloslema what is this in reference to? Does this have anything to do with the transition flickers?

**Comment by carloslema** [comment]
yup. The reason it was flickering, originally I was hiding a div after an ng-repeat. 
Before the ng-view completely rendered it had to sort out the hidding, causing the stuttering on view.

I figure, if anyone is adding anything that is doing something to the view before it finally renders it, that might the issue.

**Comment by Ionitron** [comment]
Greetings @svsool!

My sensors indicate a reply was requested, but as we have not received one yet, I am using my robot powers to close this issue. However, if you are still experiencing this issue, please feel free to reopen it by creating a new issue, and include any examples or necessary information.

Thank you for allowing me to assist you.

**Comment by Ionitron** [comment]
Greetings @svsool!

My sensors indicate a reply was requested, but as we have not received one yet, I am using my robot powers to close this issue. However, if you are still experiencing this issue, please feel free to reopen it by creating a new issue, and include any examples or necessary information.

Thank you for allowing me to assist you.

**Comment by wiltwitt** [comment]
I am also having transition issue as well.  I am using side menu and side menu will show up with previous screen on it where only menu should show.  It also flickers and show other page for 1/2 a second before it take you to the page you want.  These issues do not happen on browsers but only in ios 9.0 haven't tested on android device.

**Comment by wiltwitt** [comment]
Tested on android 4.4 and works fine.  I made a mistake saying previous page shows up on side menu panel, I meant to say previous page take up half the screen and new screen on another half.  This only seems to happen on ios

**Comment by wiltwitt** [comment]
Ok, angular-ios9-uiwebview.patch.js from Igor fixed this issue.   Thanks guys https://gist.github.com/IgorMinar/863acd413e3925bf282c

**Comment by dmarg** [comment]
@wiltwitt did you test that uiwebview patch from Igor?

**Comment by pierrybos** [comment]
+1 for angular-ios9-uiwebview.patch.js 

Saved my life

**Comment by kapil019** [comment]
@mhartington  All transition issue are resolved except. One black background is showing behind page transition.
Please suggest any option to resolve black background issue.

**Comment by danieljin** [comment]
+1 for igor's patch

**Comment by mhartington** [comment]
This should be included in 1.1.1, which was released about a week ago.

**Comment by vaidyanathanas** [comment]
when i use collection repeat below error is displayed in console

RangeError: Maximum call stack size exceeded
    at String.replace (native)
    at render (http://192.168.0.110:8100/lib/ionic/js/ionic.bundle.js:50925:12)

**Comment by pierrybos** [comment]
@vaidyanathanas can you post a codepen with code for us help you? Really you believe this problem has relation with this issue?

**Comment by vaidyanathanas** [comment]
No Sir. it was my mistake to post here. for collection-repeat i have alternative ng-repeat. but main problem is ionichistory for more than 2 views not working properly in android mobile. this is critical issue. please help me in this. thank you.

**Comment by pierrybos** [comment]
@vaidyanathanas So, please, open a new issue and do a codeshare or codepen with your problem to us can debug and help you

**Comment by vaidyanathanas** [comment]
Sir please check the below link: https://github.com/driftyco/ionic/issues/4997

**Comment by Matt2012** [comment]
I'm still getting this latest ionic 1.2.4 Horrible black flash on the right hand side of the screen ipad ios 9 during page transition.

**Comment by mlynch** [comment]
What's your app structure look like? Side menus in use at all?

**Comment by Matt2012** [comment]
https://www.dropbox.com/s/h3o1d8adly5cokx/MOV_0105.mp4?dl=0
Yes side menu project.

**Comment by Matt2012** [comment]
Looks like this fixes it:

https://forum.ionicframework.com/t/brief-black-screen-during-navigation/9480/55
(Dec 15 comment mahnuh)
ion-nav-view { background-color: transparent !important; }

**Comment by allen6131** [comment]
I'm still getting this latest ionic 1.2.4 Horrible black flash on the right hand side of the iphone 7 screen

**Comment by Maxwell2022** [comment]
Still happen in 1.3.1 on every platform (even `ionic serve`) playing with the nav, switching between screens. I'm not really sure I can reproduce it constantly

**Comment by imana97** [comment]
I tried all solutions. I still get this flickering... 
Anyone found a solution yet?
